Autocorrelation is the correlation between a variable lagged one or
more time periods and itself.

1.Use Excel to compute the autocorrelation coefficients and display the autocorrelation function for the data presented in Table 1.

In [57]:
import pandas as pd

df = pd.DataFrame (
    {
        "Month":["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],
        "VCR":[123,130,125,138,145,142,141,146,147,157,150,160]
    }
)

print(df)

   Month  VCR
0    Jan  123
1    Feb  130
2    Mar  125
3    Apr  138
4    May  145
5    Jun  142
6    Jul  141
7    Aug  146
8    Sep  147
9    Oct  157
10   Nov  150
11   Dec  160


![title](img1.png)

In [58]:
df["Yt_1"]=df["VCR"].shift(1)
df["Yt_2"]=df["VCR"].shift(2)
df["Yt_3"]=df["VCR"].shift(3)
df

Month  VCR   Yt_1   Yt_2   Yt_3
0    Jan  123    NaN    NaN    NaN
1    Feb  130  123.0    NaN    NaN
2    Mar  125  130.0  123.0    NaN
3    Apr  138  125.0  130.0  123.0
4    May  145  138.0  125.0  130.0
5    Jun  142  145.0  138.0  125.0
6    Jul  141  142.0  145.0  138.0
7    Aug  146  141.0  142.0  145.0
8    Sep  147  146.0  141.0  142.0
9    Oct  157  147.0  146.0  141.0
10   Nov  150  157.0  147.0  146.0
11   Dec  160  150.0  157.0  147.0

In [59]:
# mean Yt_k

Ybar = df["VCR"].mean()
sum_Yt = df["VCR"].sum()
mean_Yt_1 = df["Yt_1"].mean()
mean_Yt_2 = df["Yt_2"].mean()
mean_Yt_3 = df["Yt_3"].mean()

print(sum_Yt)
print(Ybar)
print(mean_Yt_1)
print(mean_Yt_2)
print(mean_Yt_3)

1704
142.0
140.36363636363637
139.4
137.44444444444446


In [60]:
df["Yt-Ybar"] = df["VCR"]-Ybar
df["Yt_1-Ybar"] = df["Yt_1"]-Ybar
df["Yt_2-Ybar"] = df["Yt_2"]-Ybar
df["Yt_3-Ybar"] = df["Yt_3"]-Ybar

df

Month  VCR   Yt_1   Yt_2   Yt_3  Yt-Ybar  Yt_1-Ybar  Yt_2-Ybar  Yt_3-Ybar
0    Jan  123    NaN    NaN    NaN    -19.0        NaN        NaN        NaN
1    Feb  130  123.0    NaN    NaN    -12.0      -19.0        NaN        NaN
2    Mar  125  130.0  123.0    NaN    -17.0      -12.0      -19.0        NaN
3    Apr  138  125.0  130.0  123.0     -4.0      -17.0      -12.0      -19.0
4    May  145  138.0  125.0  130.0      3.0       -4.0      -17.0      -12.0
5    Jun  142  145.0  138.0  125.0      0.0        3.0       -4.0      -17.0
6    Jul  141  142.0  145.0  138.0     -1.0        0.0        3.0       -4.0
7    Aug  146  141.0  142.0  145.0      4.0       -1.0        0.0        3.0
8    Sep  147  146.0  141.0  142.0      5.0        4.0       -1.0        0.0
9    Oct  157  147.0  146.0  141.0     15.0        5.0        4.0       -1.0
10   Nov  150  157.0  147.0  146.0      8.0       15.0        5.0        4.0
11   Dec  160  150.0  157.0  147.0     18.0        8.0       15.0        5.0

In [64]:
df["(Yt-Ybar)(Yt_1-Ybar)"]=df["Yt-Ybar"]*df["Yt_1-Ybar"]
df["(Yt-Ybar)(Yt_2-Ybar)"]=df["Yt-Ybar"]*df["Yt_2-Ybar"]
df["(Yt-Ybar)(Yt_3-Ybar)"]=df["Yt-Ybar"]*df["Yt_3-Ybar"]
df["(Yt-Ybar)^2"]=df["Yt-Ybar"]**2
df

Month  VCR   Yt_1   Yt_2   Yt_3  Yt-Ybar  Yt_1-Ybar  Yt_2-Ybar  Yt_3-Ybar  \
0    Jan  123    NaN    NaN    NaN    -19.0        NaN        NaN        NaN   
1    Feb  130  123.0    NaN    NaN    -12.0      -19.0        NaN        NaN   
2    Mar  125  130.0  123.0    NaN    -17.0      -12.0      -19.0        NaN   
3    Apr  138  125.0  130.0  123.0     -4.0      -17.0      -12.0      -19.0   
4    May  145  138.0  125.0  130.0      3.0       -4.0      -17.0      -12.0   
5    Jun  142  145.0  138.0  125.0      0.0        3.0       -4.0      -17.0   
6    Jul  141  142.0  145.0  138.0     -1.0        0.0        3.0       -4.0   
7    Aug  146  141.0  142.0  145.0      4.0       -1.0        0.0        3.0   
8    Sep  147  146.0  141.0  142.0      5.0        4.0       -1.0        0.0   
9    Oct  157  147.0  146.0  141.0     15.0        5.0        4.0       -1.0   
10   Nov  150  157.0  147.0  146.0      8.0       15.0        5.0        4.0   
11   Dec  160  150.0  157.0  147.0     18.0        8.0       15.0        5.0   

    (Yt-Ybar)(Yt_1-Ybar)  (Yt-Ybar)(Yt_2-Ybar)  (Yt-Ybar)(Yt_3-Ybar)  \
0                    NaN                   NaN                   NaN   
1                  228.0                   NaN                   NaN   
2                  204.0                 323.0                   NaN   
3                   68.0                  48.0                  76.0   
4                  -12.0                 -51.0                 -36.0   
5                    0.0                  -0.0                  -0.0   
6                   -0.0                  -3.0                   4.0   
7                   -4.0                   0.0                  12.0   
8                   20.0                  -5.0                   0.0   
9                   75.0                  60.0                 -15.0   
10                 120.0                  40.0                  32.0   
11                 144.0                 270.0                  90.0   

    (Yt-Ybar)^2  
0         361.0  
1         144.0  
2         289.0  
3          16.0  
4           9.0  
5           0.0  
6           1.0  
7          16.0  
8          25.0  
9         225.0  
10         64.0  
11        324.0

In [68]:
r1=df["(Yt-Ybar)(Yt_1-Ybar)"].sum()/df["(Yt-Ybar)^2"].sum()
r2=df["(Yt-Ybar)(Yt_2-Ybar)"].sum()/df["(Yt-Ybar)^2"].sum()
r3=df["(Yt-Ybar)(Yt_3-Ybar)"].sum()/df["(Yt-Ybar)^2"].sum()
print(r1)
print(r2)
print(r3)

0.5719131614654003
0.4626865671641791
0.11058344640434192


Results 

Positive lag 1 autocorrelation exists in this time series. The correlation between Yt and Yt-1
or the autocorrelation for time lag 1, is .572. This means that the successive monthly sales of VCRs are somewhat correlated with each other.

It appears that moderate autocorrelation exists in this time series lagged two time periods. The correlation between Yt and Yt-2
, or the autocorrelation for time lag 2, is .463.

The autocorrelation for time lag 3, is .111. This means a weak autocorrelation exists in this
time series lagged three time periods.

Notice that the autocorrelation coefficient at time lag 2 (.463) is less than the
autocorrelation coefficient at time lag 1 (.572). Generally, as the number of time lags ( k )
increases, the magnitudes of the autocorrelation coefficients decrease.